src/codegen/generate.ts

```diff
@@ -385,85 +375,67 @@ export default class ApiGenerator {
-      // Create a type if no readOnly or writeOnly properties found
-      if (!this.isSchemaReadOnly(schema) && !this.isSchemaWriteOnly(schema)) {
-        this.aliases.push(
-          cg.createTypeAliasDeclaration({
-            modifiers: [cg.modifier.export],
-            name: alias,
-            type,
-          }),
-        );
-      } else {
-        // Create an interface to be extended
-        this.aliases.push(
-          cg.createIntefaceAliasDeclaration({
-            modifiers: [cg.modifier.export],
-            name: alias,
-            type,
-          }),
-        );
-      }
-    }
-    if (onlyMode === "readOnly" && !this.refs[$ref]["readOnly"]) {
-      const schema = this.resolve<SchemaObject>(obj);
-      const name = schema.title || getRefName($ref);
-      if (this.isSchemaReadOnly(schema)) {
+      if (this.checkSchemaOnlyMode(schema, "readOnly")) {
         const readOnlyAlias = this.getUniqueAlias(
           toIdentifier(name, true, "readOnly"),
         );
-        ref = this.refs[$ref]["readOnly"] = factory.createTypeReferenceNode(
+        this.refs[$ref]["readOnly"] = factory.createTypeReferenceNode(
           readOnlyAlias,
           undefined,
         );
 
         const readOnlyType = this.getTypeFromSchema(schema, name, "readOnly");
         this.aliases.push(
-          cg.createIntefaceAliasDeclaration({
+          cg.createTypeAliasDeclaration({
             modifiers: [cg.modifier.export],
             name: readOnlyAlias,
             type: readOnlyType,
-            inheritedNodeNames: [name],
           }),
         );
       }
-    }
-    if (onlyMode === "writeOnly" && !this.refs[$ref]["writeOnly"]) {
-      const schema = this.resolve<SchemaObject>(obj);
-      const name = schema.title || getRefName($ref);
-      if (this.isSchemaWriteOnly(schema)) {
+
+      if (this.checkSchemaOnlyMode(schema, "writeOnly")) {
         const writeOnlyAlias = this.getUniqueAlias(
           toIdentifier(name, true, "writeOnly"),
         );
-        ref = this.refs[$ref]["writeOnly"] = factory.createTypeReferenceNode(
+        this.refs[$ref]["writeOnly"] = factory.createTypeReferenceNode(
           writeOnlyAlias,
           undefined,
         );
         const writeOnlyType = this.getTypeFromSchema(schema, name, "writeOnly");
         this.aliases.push(
-          cg.createIntefaceAliasDeclaration({
+          cg.createTypeAliasDeclaration({
             modifiers: [cg.modifier.export],
             name: writeOnlyAlias,
             type: writeOnlyType,
-            inheritedNodeNames: [name],
           }),
         );
       }
     }
 
     // If not ref fallback to the regular reference
-    return ref ?? this.refs[$ref]["none"];
+    return this.refs[$ref][onlyMode || "base"] ?? this.refs[$ref].base;
   }
 
   getUnionType(
     variants: (OpenAPIV3.ReferenceObject | SchemaObject)[],
     discriminator?: OpenAPIV3.DiscriminatorObject,
+    onlyMode?: OnlyMode,
```

Comment:
the `onlyMode` is now passed around much more and more deeply so that when we create a Parent type in "writeOnly" mode, and that one has a $ref for example, we'll try attaching the writeOnly version of that ref.